### PR TITLE
feat: implement sind intrinsic function

### PR DESCRIPTION
**Pull request by parth121101**

### CODE REVIEW

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Left comment above

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -1073,6 +1073,18 @@ LFORTRAN_API double _lfortran_dsin(double x)
...
+{
+    float radians = (x * PI) / 180.0;
+    return sin(radians);
+}
+
+LFORTRAN_API double _lfortran_dsind(double x)
+{
+    double radians = (x * PI) / 180.0;
+    return sin(radians);
+}
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1086]: If we are going to use a macro style, wouldn't it make sense to instantiate this operation in ASR itself? That way we don't need to ship these simple runtime functions.

**Comment by kargl** [comment]
As one of the individuals that worked on gfortran's implementation of the degree trig functions, I'll politely suggest that one might want to reconsider the implementation details.  Consider,

```
program foo
   implicit none
   integer n
   real x, y, z
   real, parameter :: pi = 4 * atan(1.)
   do n = 0, 10
      x = n * 360 + 30
      y = sind(x)
      z = sin((x * pi) / 180)
      print '(2(F10.7,1X,Z8.8,1X))', y, y, z, z
   end do
end program foo
```
    % gfcx -o z a.f90 && ./z
     0.5000000 3F000000  0.5000000 3F000000
     0.5000000 3F000000  0.5000005 3F000008
     0.5000000 3F000000  0.5000002 3F000004
     0.5000000 3F000000  0.5000008 3F00000D
     0.5000000 3F000000  0.4999997 3EFFFFF6
     0.5000000 3F000000  0.5000002 3F000004
     0.5000000 3F000000  0.4999992 3EFFFFE4
     0.5000000 3F000000  0.5000014 3F000017
     0.5000000 3F000000  0.5000036 3F00003C
     0.5000000 3F000000  0.4999992 3EFFFFE5
     0.5000000 3F000000  0.5000014 3F000018

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -3226,6 +3227,51 @@ namespace Asind {
...
+            fn_symtab->add_symbol(c_func_name, s);
+            dep.push_back(al, s2c(al, c_func_name));
+            body.push_back(al, b.Assignment(result, b.Call(s, args, return_type)));
+        }
+
+        ASR::symbol_t *new_symbol = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        scope->add_symbol(fn_name, new_symbol);
+        return b.Call(new_symbol, new_args, return_type);
+    }
```
[the remark this answers is not in the document]
Comment (parth121101) [reply src/libasr/pass/intrinsic_functions.h:3261]: @Pranavchiku, I tried creating the macro for this kind of function. The results that the function returns does not match with the gfortran. Can you please review it and let me know the problem?

File: src/libasr/asr_builder.h [hunk 3]
```diff
@@ -1064,6 +1064,22 @@ class ASRBuilder {
...
+    ASR::expr_t* math_helper_func2(ASR::expr_t* result, ASRUtils::ASRBuilder b, SymbolTable *fn_symtab, SetChar dep, Vec<ASR::stmt_t*> body, Vec<ASR::expr_t*> args, std::string c_func_name, ASR::ttype_t* return_type, int n_args, Vec<ASR::ttype_t*>& arg_types) {
+    
+        ASR::symbol_t *s = b.create_c_func(c_func_name, fn_symtab, return_type, n_args, arg_types);
+        fn_symtab->add_symbol(c_func_name, s);
+        dep.push_back(al, s2c(al, c_func_name));
+        body.push_back(al, b.Assignment(result, b.Call(s, args, return_type)));
+        
+
+        return result;
+    }
```
Comment (Pranavchiku) [inline src/libasr/asr_builder.h:1082]: This is not what we want, you need to create something similar to `create_trig`, you may refer it.

**Comment by Pranavchiku** [comment]
> As one of the individuals that worked on gfortran's implementation of the degree trig functions, I'll politely suggest that one might want to reconsider the implementation details. Consider,
> 
> ```
> program foo
>    implicit none
>    integer n
>    real x, y, z
>    real, parameter :: pi = 4 * atan(1.)
>    do n = 0, 10
>       x = n * 360 + 30
>       y = sind(x)
>       z = sin((x * pi) / 180)
>       print '(2(F10.7,1X,Z8.8,1X))', y, y, z, z
>    end do
> end program foo
> ```
> 
>  ```
> % gfcx -o z a.f90 && ./z
>  0.5000000 3F000000  0.5000000 3F000000
>  0.5000000 3F000000  0.5000005 3F000008
>  0.5000000 3F000000  0.5000002 3F000004
>  0.5000000 3F000000  0.5000008 3F00000D
>  0.5000000 3F000000  0.4999997 3EFFFFF6
>  0.5000000 3F000000  0.5000002 3F000004
>  0.5000000 3F000000  0.4999992 3EFFFFE4
>  0.5000000 3F000000  0.5000014 3F000017
>  0.5000000 3F000000  0.5000036 3F00003C
>  0.5000000 3F000000  0.4999992 3EFFFFE5
>  0.5000000 3F000000  0.5000014 3F000018
> ```

We expected divergence at higher precision but this is something concerning ( 1e-7 ). I'll be happy to get your input on how shall we proceed to implement this and resolve this divergence.

We faced something similar at #3894.

**Comment by kargl** [comment]
> > As one of the individuals that worked on gfortran's implementation of the degree trig functions, I'll politely suggest that one might want to reconsider the implementation details. Consider,

(program elided)

> > ```
> > % gfcx -o z a.f90 && ./z
> > 0.5000000 3F000000  0.5000000 3F000000
>> (elided)
> > 0.5000000 3F000000  0.5000014 3F000018
> > ```
> 
> We expected divergence at higher precision but this is something concerning ( 1e-7 ). I'll be happy to get your input on how shall we proceed to implement this and resolve this divergence.
> 
> We faced something similar at #3894.

Within `sind(n*360 + 30)`, the result is exactly 0.5.  In principle, no exception should be raised, but I don't remember if we worried about that detail with gfortran.   FE_INEXACT may have occurred.   There are other special cases such as `sind(n*360+90) = 1` exactly.  Similar statements hold for the other degree trig functions.

First, one needs to use reflection symmetry so `s = x > 0 ? 1:-1; ax = abs(x)`.  Before returning the computed value, you look at `s` to adjust the sign.   This guarantees reflection symmetry holds in the face of floating pointing arithmetic.

What then needs to be done is argument reduction, $ax = n * 360 + r$, with $r$ in [0,360).  Now, you work with `r`.   First, check for special cases, then fold `r` in [0,360) into `a` in [0,45] with appropriate trig identities.  The last step is the conversion of `a` into radians, but you cannot use `b = a * pi / 180` as rounding issues can arise.  First, compute `pi / 180 = hi + lo` with `hi` the leading `digits(x)/2`  bits and  `lo` is `digits(x)` trailing bits.  So, `b = a*lo + b*hi`

**Comment by Pranavchiku** [comment]
I opened a dedicated issue https://github.com/lfortran/lfortran/issues/3912 for this.

**Comment by Pranavchiku** [comment]
This PR can be closed with reference to #3914

**Pull request closed without merging**